### PR TITLE
Add latest commit's SHA to content response

### LINKED ISSUE #12840: API: please make commit hash/timestamp available with ​/repos​/{owner}​/{repo}​/contents​/{filepath}
- Gitea version (or commit ref): try.gitea.org & codeberg.org
- Git version:
- Operating system:
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL: https://try.gitea.io/magnusja/libaums/src/branch/develop/app/build.gradle)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
I'm trying to fetch commit timestamps for a given file via the API. Using `/repos​/{owner}​/{repo}​/contents​/{filepath}` (eg `https://try.gitea.io/api/v1/repos/magnusja/libaums/contents/app%2Fbuild.gradle`) reports details for the file quite fine, but obviously lists a wrong sha: using the sha given there as `{ref}` for `​/repos​/{owner}​/{repo}​/commits​/{ref}​/statuses` gives an empty result, for `https://try.gitea.io/api/v1/repos/magnusja/libaums/commits?sha=049bdf7bbc8ff39411b618ff855a14ed3446dcc1&limit=1` it even results in a 503 server error. Trying the sha on `​/repos​/{owner}​/{repo}​/git​/trees​/{sha}` (where it explicitly says "sha") results in a "bad request: sha not found".

Reproducible for other repos/files on eg Codeberg.org – I could not find a single valid commit hash in the values given for `sha`, and no way to get the correct commit hashes.

**UPDATE:** as [pointed out](#issuecomment-700696627) by @mrsdizzie: "this is the file SHA and not a bug" – hence converting this from a "bug report" to a feature request: As currently there's no way to get the file's timestamp/commit-hash, please add it at this level (https://try.gitea.io/api/swagger#/repository/repoGetContents – as suggested by @mrsdizzie). Ideally both. More details on the commit itself could be obtained easily using ​/repos​/{owner}​/{repo}​/commits/{ref}​/statuses, for example, once the commit hash is known.

### PR DESCRIPTION
**Pull request by Gusted**
- When requesting the contents of a filepath, add the latest commit's SHA to the requested file.
- Resolves #12840

### CODE REVIEW

**Comment by zeripath** [comment]
I wonder if we should be hooking into the lastcommit cache here, but I guess the issue is we don't really want the cache to explode however, I'm going to assume that most of the time the user here is going to keep requesting the same object over and over to see if the last commit has changed which is eminently useful for caching.

The real problem with this underlying feature request is that it totally ignores the actual cost of finding out this information. It's very much not free on large repositories and can require significant calculation.

**Comment by zeripath** [comment]
Thinking on - we absolutely should be hooking in to the last commit cache here.

We may also need to consider if this information needs to be only for the default branch or for those who are logged in.

**Comment by zeripath** [comment]
I think we need #20444 before this is merged.

**Comment by Gusted** [comment]
Since #20444 is merged. Do you have a strong opinion on the matter when we should return this information? I'm fine either way, as we don't *really* have use-case to go on, other than https://github.com/go-gitea/gitea/issues/12840 which only uses default branch.

**Comment by zeripath** [comment]
I'm pretty certain that gitRepos you're using here won't have had a LastCommitCache set because that is only set once we know a reference commit to provide the reasoning for caching. (Yes I know that means that the LCC should be on the git.Commit not the git.Repo - that would mean that git.Repo should have a commit cache part - as they kinda do in the gogit variant - and then the LCC just wraps the calls to the system cache before asking the repo to get commits - which it can choose to get from its cache.)


You need to assign the lastcommitcache as per utils.ResolveRefOrSha

```go
	if ctx.Repo.GitRepo != nil && ctx.Repo.GitRepo.LastCommitCache == nil {
		commitsCount, err := cache.GetInt64(ctx.Repo.Repository.GetCommitsCountCacheKey(ref, true), func() (int64, error) {
			commit, err := ctx.Repo.GitRepo.GetCommit(sha)
			if err != nil {
				return 0, err
			}
			return commit.CommitsCount()
		})
		if err != nil {
			log.Error("Unable to get commits count for %s in %s. Error: %v", sha, ctx.Repo.Repository.FullName(), err)
			return sha
		}
		ctx.Repo.GitRepo.LastCommitCache = git.NewLastCommitCache(commitsCount, ctx.Repo.Repository.FullName(), ctx.Repo.GitRepo, cache.GetCache())
	}
```

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
Build failures were due to the fact that this function needs to be shared across both gogit and nogogit variants.

I've fixed that.

**Review by wxiaoguang — APPROVED** [review]
LGTM

(curious about why the `NewLastCommitCache(count ...)` must have the count argument. It seems that `NewLastCommitCache` may be possible to handle the count internally?)

File: routers/api/v1/utils/git.go [hunk 1]
```diff
@@ -35,19 +34,9 @@ func ResolveRefOrSha(ctx *context.APIContext, ref string) string {
...
-				return 0, err
-			}
-			return commit.CommitsCount()
-		})
-		if err != nil {
-			log.Error("Unable to get commits count for %s in %s. Error: %v", sha, ctx.Repo.Repository.FullName(), err)
-			return sha
-		}
-		ctx.Repo.GitRepo.LastCommitCache = git.NewLastCommitCache(commitsCount, ctx.Repo.Repository.FullName(), ctx.Repo.GitRepo, cache.GetCache())
+	err := ctx.Repo.GitRepo.AddLastCommitCache(ctx.Repo.Repository.GetCommitsCountCacheKey(ref, ref != sha), ctx.Repo.Repository.FullName(), sha)
```
Comment (lunny) [inline routers/api/v1/utils/git.go]: ctx.Repo.GitRepo != nil

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**